File: clients/android/app/src/main/java/local/visual/corepose/PersonRoi.kt
package local.visual.corepose

data class PixelBox(
    val x1: Double,
    val y1: Double,
    val x2: Double,
    val y2: Double,
) {
    fun toBBox(): BBox = BBox(x1.toFloat(), y1.toFloat(), x2.toFloat(), y2.toFloat())
}

data class CropSpec(
    val ox: Int,
    val oy: Int,
    val x2: Int,
    val y2: Int,
    val scale: Double,
)

object PersonRoi {
    const val HIST_H_BINS = 30
    const val HIST_S_BINS = 32
    const val MIN_SAT = 40
    const val MAX_VAL = 250
    const val SEARCH_EXPAND = 1.8
    const val POSE_PAD = 0.35
    const val MIN_CROP_SIDE = 256
    const val TERM_COUNT = 12
    const val TERM_EPS = 1.0

    fun denormBox(seed: NormBox, width: Int, height: Int): PixelBox {
        return PixelBox(
            seed.x1 * width.toDouble(),
            seed.y1 * height.toDouble(),
            seed.x2 * width.toDouble(),
            seed.y2 * height.toDouble(),
        )
    }

    fun clipBox(box: PixelBox, width: Int, height: Int): IntArray {
        val nx1 = maxOf(0.0, minOf(box.x1, box.x2)).toInt()
        val ny1 = maxOf(0.0, minOf(box.y1, box.y2)).toInt()
        val nx2 = minOf(width.toDouble(), maxOf(box.x1, box.x2)).toInt()
        val ny2 = minOf(height.toDouble(), maxOf(box.y1, box.y2)).toInt()
        if (nx2 <= nx1 || ny2 <= ny1) {
            return intArrayOf(0, 0, width, height)
        }
        return intArrayOf(nx1, ny1, nx2, ny2)
    }

    fun cropRect(
        box: PixelBox,
        width: Int,
        height: Int,
        pad: Double = POSE_PAD,
        minSide: Int = MIN_CROP_SIDE,
    ): CropSpec {
        val bw = maxOf(1.0, box.x2 - box.x1)
        val bh = maxOf(1.0, box.y2 - box.y1)
        val padded = PixelBox(
            box.x1 - bw * pad,
            box.y1 - bh * pad,
            box.x2 + bw * pad,
            box.y2 + bh * pad,
        )
        val clipped = clipBox(padded, width, height)
        val cw = (clipped[2] - clipped[0]).coerceAtLeast(1)
        val ch = (clipped[3] - clipped[1]).coerceAtLeast(1)
        val shortest = minOf(cw, ch)
        val scale = if (shortest < minSide) minSide.toDouble() / shortest.toDouble() else 1.0
        return CropSpec(clipped[0], clipped[1], clipped[2], clipped[3], scale)
    }

    fun remapPoint(x: Double, y: Double, ox: Int, oy: Int, scale: Double): Pair<Double, Double> {
        val inv = if (scale == 0.0) 1.0 else 1.0 / scale
        return x * inv + ox to y * inv + oy
    }

    fun buildHist(pixels: IntArray, width: Int, height: Int, box: PixelBox): FloatArray {
        val hist = FloatArray(HIST_H_BINS * HIST_S_BINS)
        val clip = clipBox(box, width, height)
        val hsv = FloatArray(3)
        for (y in clip[1] until clip[3]) {
            for (x in clip[0] until clip[2]) {
                val argb = pixels[y * width + x]
                toHsv(argb, hsv)
                val h = hsv[0]
                val s = hsv[1]
                val v = hsv[2]
                if (s < MIN_SAT || v < 32 || v > MAX_VAL) {
                    continue
                }
                val hb = ((h / 180.0) * HIST_H_BINS).toInt().coerceIn(0, HIST_H_BINS - 1)
                val sb = ((s / 256.0) * HIST_S_BINS).toInt().coerceIn(0, HIST_S_BINS - 1)
                hist[hb * HIST_S_BINS + sb] += 1f
            }
        }
        var max = 0f
        for (value in hist) {
            if (value > max) {
                max = value
            }
        }
        if (max > 0f) {
            val gain = 255f / max
            for (i in hist.indices) {
                hist[i] *= gain
            }
        }
        return hist
    }

    fun search(
        pixels: IntArray,
        width: Int,
        height: Int,
        hist: FloatArray,
        prev: PixelBox,
    ): PixelBox {
        val back = backProject(pixels, width, height, hist)
        val window = expandWindow(prev, width, height, SEARCH_EXPAND)
        var x = window[0]
        var y = window[1]
        var w = window[2]
        var h = window[3]
        for (iter in 0 until TERM_COUNT) {
            var m00 = 0.0
            var m10 = 0.0
            var m01 = 0.0
            val x2 = (x + w).coerceAtMost(width)
            val y2 = (y + h).coerceAtMost(height)
            for (py in y until y2) {
                val row = py * width
                for (px in x until x2) {
                    val mass = back[row + px].toDouble()
                    m00 += mass
                    m10 += px * mass
                    m01 += py * mass
                }
            }
            if (m00 < 1.0) {
                break
            }
            val cx = (m10 / m00).toInt()
            val cy = (m01 / m00).toInt()
            val nx = (cx - w / 2).coerceIn(0, width - 1)
            val ny = (cy - h / 2).coerceIn(0, height - 1)
            val dx = kotlin.math.abs(nx - x)
            val dy = kotlin.math.abs(ny - y)
            x = nx
            y = ny
            if (dx < TERM_EPS && dy < TERM_EPS) {
                break
            }
        }
        if (w >= 8 && h >= 8) {
            return PixelBox(
                x.toDouble(),
                y.toDouble(),
                (x + w).coerceAtMost(width).toDouble(),
                (y + h).coerceAtMost(height).toDouble(),
            )
        }
        return peakFallback(back, width, height, prev)
    }

    fun blendHist(prev: FloatArray, next: FloatArray, alpha: Float = 0.15f): FloatArray {
        val mixed = FloatArray(prev.size)
        var max = 0f
        for (i in prev.indices) {
            val value = (1f - alpha) * prev[i] + alpha * next[i]
            mixed[i] = value
            if (value > max) {
                max = value
            }
        }
        if (max > 0f) {
            val gain = 255f / max
            for (i in mixed.indices) {
                mixed[i] *= gain
            }
        }
        return mixed
    }

    private fun backProject(pixels: IntArray, width: Int, height: Int, hist: FloatArray): FloatArray {
        val out = FloatArray(pixels.size)
        val hsv = FloatArray(3)
        for (i in pixels.indices) {
            toHsv(pixels[i], hsv)
            val hb = ((hsv[0] / 180.0) * HIST_H_BINS).toInt().coerceIn(0, HIST_H_BINS - 1)
            val sb = ((hsv[1] / 256.0) * HIST_S_BINS).toInt().coerceIn(0, HIST_S_BINS - 1)
            out[i] = hist[hb * HIST_S_BINS + sb]
        }
        return out
    }

    private fun expandWindow(box: PixelBox, width: Int, height: Int, scale: Double): IntArray {
        val cx = 0.5 * (box.x1 + box.x2)
        val cy = 0.5 * (box.y1 + box.y2)
        val bw = maxOf(8.0, (box.x2 - box.x1) * scale)
        val bh = maxOf(8.0, (box.y2 - box.y1) * scale)
        val nx1 = maxOf(0.0, cx - bw / 2.0).toInt()
        val ny1 = maxOf(0.0, cy - bh / 2.0).toInt()
        val nw = minOf(width - nx1, bw.toInt()).coerceAtLeast(1)
        val nh = minOf(height - ny1, bh.toInt()).coerceAtLeast(1)
        return intArrayOf(nx1, ny1, nw, nh)
    }

    private fun peakFallback(back: FloatArray, width: Int, height: Int, prev: PixelBox): PixelBox {
        val sw = maxOf(1, width / 4)
        val sh = maxOf(1, height / 4)
        var best = -1f
        var px = width / 2
        var py = height / 2
        for (sy in 0 until sh) {
            for (sx in 0 until sw) {
                val value = back[(sy * 4).coerceAtMost(height - 1) * width + (sx * 4).coerceAtMost(width - 1)]
                if (value > best) {
                    best = value
                    px = sx * 4
                    py = sy * 4
                }
            }
        }
        val bw = maxOf(16.0, prev.x2 - prev.x1)
        val bh = maxOf(16.0, prev.y2 - prev.y1)
        return PixelBox(
            maxOf(0.0, px - bw / 2),
            maxOf(0.0, py - bh / 2),
            minOf(width.toDouble(), px + bw / 2),
            minOf(height.toDouble(), py + bh / 2),
        )
    }

    private fun toHsv(argb: Int, out: FloatArray) {
        val r = (argb shr 16) and 0xFF
        val g = (argb shr 8) and 0xFF
        val b = argb and 0xFF
        val max = maxOf(r, g, b)
        val min = minOf(r, g, b)
        val delta = max - min
        val v = max.toFloat()
        val s = if (max == 0) 0f else delta * 255f / max
        val h = if (delta == 0) {
            0f
        } else if (max == r) {
            60f * (((g - b).toFloat() / delta + 6f) % 6f)
        } else if (max == g) {
            60f * ((b - r).toFloat() / delta + 2f)
        } else {
            60f * ((r - g).toFloat() / delta + 4f)
        }
        out[0] = (h / 2f).coerceIn(0f, 179f)
        out[1] = s.coerceIn(0f, 255f)
        out[2] = v
    }
}

File: clients/ios/VisualPose/PersonRoi.swift
import Foundation

struct PixelBox {
    var x1: Double
    var y1: Double
    var x2: Double
    var y2: Double

    func toBBox() -> BBox {
        BBox(x1: Float(x1), y1: Float(y1), x2: Float(x2), y2: Float(y2))
    }
}

struct CropSpec {
    var ox: Int
    var oy: Int
    var x2: Int
    var y2: Int
    var scale: Double
}

enum PersonRoi {
    static let histHBins = 30
    static let histSBins = 32
    static let minSat: Float = 40
    static let maxVal: Float = 250
    static let searchExpand = 1.8
    static let posePad = 0.35
    static let minCropSide = 256
    static let termCount = 12
    static let termEps = 1.0

    static func denormBox(_ seed: NormBox, width: Int, height: Int) -> PixelBox {
        PixelBox(
            x1: Double(seed.x1) * Double(width),
            y1: Double(seed.y1) * Double(height),
            x2: Double(seed.x2) * Double(width),
            y2: Double(seed.y2) * Double(height)
        )
    }

    static func clipBox(_ box: PixelBox, width: Int, height: Int) -> [Int] {
        let nx1 = Int(max(0.0, min(box.x1, box.x2)))
        let ny1 = Int(max(0.0, min(box.y1, box.y2)))
        let nx2 = Int(min(Double(width), max(box.x1, box.x2)))
        let ny2 = Int(min(Double(height), max(box.y1, box.y2)))
        if nx2 <= nx1 || ny2 <= ny1 {
            return [0, 0, width, height]
        }
        return [nx1, ny1, nx2, ny2]
    }

    static func cropRect(
        _ box: PixelBox,
        width: Int,
        height: Int,
        pad: Double = posePad,
        minSide: Int = minCropSide
    ) -> CropSpec {
        let bw = max(1.0, box.x2 - box.x1)
        let bh = max(1.0, box.y2 - box.y1)
        let padded = PixelBox(
            x1: box.x1 - bw * pad,
            y1: box.y1 - bh * pad,
            x2: box.x2 + bw * pad,
            y2: box.y2 + bh * pad
        )
        let clipped = clipBox(padded, width: width, height: height)
        let cw = max(clipped[2] - clipped[0], 1)
        let ch = max(clipped[3] - clipped[1], 1)
        let shortest = min(cw, ch)
        let scale = shortest < minSide ? Double(minSide) / Double(shortest) : 1.0
        return CropSpec(ox: clipped[0], oy: clipped[1], x2: clipped[2], y2: clipped[3], scale: scale)
    }

    static func remapPoint(x: Double, y: Double, ox: Int, oy: Int, scale: Double) -> (Double, Double) {
        let inv = scale == 0 ? 1.0 : 1.0 / scale
        return (x * inv + Double(ox), y * inv + Double(oy))
    }

    static func buildHist(pixels: [UInt32], width: Int, height: Int, box: PixelBox) -> [Float] {
        var hist = [Float](repeating: 0, count: histHBins * histSBins)
        let clip = clipBox(box, width: width, height: height)
        var hsv: [Float] = [0, 0, 0]
        for y in clip[1]..<clip[3] {
            for x in clip[0]..<clip[2] {
                toHsv(pixels[y * width + x], &hsv)
                let h = hsv[0]
                let s = hsv[1]
                let v = hsv[2]
                if s < minSat || v < 32 || v > maxVal { continue }
                let hb = min(max(Int((Double(h) / 180.0) * Double(histHBins)), 0), histHBins - 1)
                let sb = min(max(Int((Double(s) / 256.0) * Double(histSBins)), 0), histSBins - 1)
                hist[hb * histSBins + sb] += 1
            }
        }
        var maxv: Float = 0
        for value in hist where value > maxv {
            maxv = value
        }
        if maxv > 0 {
            let gain = 255 / maxv
            for i in hist.indices {
                hist[i] *= gain
            }
        }
        return hist
    }

    static func search(
        pixels: [UInt32],
        width: Int,
        height: Int,
        hist: [Float],
        prev: PixelBox
    ) -> PixelBox {
        let back = backProject(pixels: pixels, width: width, height: height, hist: hist)
        let window = expandWindow(prev, width: width, height: height, scale: searchExpand)
        var x = window[0]
        var y = window[1]
        var w = window[2]
        var h = window[3]
        for _ in 0..<termCount {
            var m00 = 0.0
            var m10 = 0.0
            var m01 = 0.0
            let x2 = min(x + w, width)
            let y2 = min(y + h, height)
            if y < y2 && x < x2 {
                for py in y..<y2 {
                    let row = py * width
                    for px in x..<x2 {
                        let mass = Double(back[row + px])
                        m00 += mass
                        m10 += Double(px) * mass
                        m01 += Double(py) * mass
                    }
                }
            }
            if m00 < 1.0 { break }
            let cx = Int(m10 / m00)
            let cy = Int(m01 / m00)
            let nx = min(max(cx - w / 2, 0), width - 1)
            let ny = min(max(cy - h / 2, 0), height - 1)
            let dx = abs(nx - x)
            let dy = abs(ny - y)
            x = nx
            y = ny
            if Double(dx) < termEps && Double(dy) < termEps { break }
        }
        if w >= 8 && h >= 8 {
            return PixelBox(
                x1: Double(x),
                y1: Double(y),
                x2: Double(min(x + w, width)),
                y2: Double(min(y + h, height))
            )
        }
        return peakFallback(back, width: width, height: height, prev: prev)
    }

    static func blendHist(_ prev: [Float], _ next: [Float], alpha: Float = 0.15) -> [Float] {
        var mixed = [Float](repeating: 0, count: prev.count)
        var maxv: Float = 0
        for i in prev.indices {
            let value = (1 - alpha) * prev[i] + alpha * next[i]
            mixed[i] = value
            if value > maxv { maxv = value }
        }
        if maxv > 0 {
            let gain = 255 / maxv
            for i in mixed.indices {
                mixed[i] *= gain
            }
        }
        return mixed
    }

    private static func backProject(pixels: [UInt32], width: Int, height: Int, hist: [Float]) -> [Float] {
        var out = [Float](repeating: 0, count: pixels.count)
        var hsv: [Float] = [0, 0, 0]
        for i in pixels.indices {
            toHsv(pixels[i], &hsv)
            let hb = min(max(Int((Double(hsv[0]) / 180.0) * Double(histHBins)), 0), histHBins - 1)
            let sb = min(max(Int((Double(hsv[1]) / 256.0) * Double(histSBins)), 0), histSBins - 1)
            out[i] = hist[hb * histSBins + sb]
        }
        return out
    }

    private static func expandWindow(_ box: PixelBox, width: Int, height: Int, scale: Double) -> [Int] {
        let cx = 0.5 * (box.x1 + box.x2)
        let cy = 0.5 * (box.y1 + box.y2)
        let bw = max(8.0, (box.x2 - box.x1) * scale)
        let bh = max(8.0, (box.y2 - box.y1) * scale)
        let nx1 = Int(max(0.0, cx - bw / 2.0))
        let ny1 = Int(max(0.0, cy - bh / 2.0))
        let nw = max(min(width - nx1, Int(bw)), 1)
        let nh = max(min(height - ny1, Int(bh)), 1)
        return [nx1, ny1, nw, nh]
    }

    private static func peakFallback(_ back: [Float], width: Int, height: Int, prev: PixelBox) -> PixelBox {
        let sw = max(1, width / 4)
        let sh = max(1, height / 4)
        var best: Float = -1
        var px = width / 2
        var py = height / 2
        for sy in 0..<sh {
            for sx in 0..<sw {
                let value = back[min(sy * 4, height - 1) * width + min(sx * 4, width - 1)]
                if value > best {
                    best = value
                    px = sx * 4
                    py = sy * 4
                }
            }
        }
        let bw = max(16.0, prev.x2 - prev.x1)
        let bh = max(16.0, prev.y2 - prev.y1)
        return PixelBox(
            x1: max(0.0, Double(px) - bw / 2),
            y1: max(0.0, Double(py) - bh / 2),
            x2: min(Double(width), Double(px) + bw / 2),
            y2: min(Double(height), Double(py) + bh / 2)
        )
    }

    private static func toHsv(_ argb: UInt32, _ out: inout [Float]) {
        let r = Int((argb >> 16) & 0xFF)
        let g = Int((argb >> 8) & 0xFF)
        let b = Int(argb & 0xFF)
        let maxc = max(r, max(g, b))
        let minc = min(r, min(g, b))
        let delta = maxc - minc
        let v = Float(maxc)
        let s: Float = maxc == 0 ? 0 : Float(delta) * 255 / Float(maxc)
        let h: Float
        if delta == 0 {
            h = 0
        } else if maxc == r {
            h = 60 * ((Float(g - b) / Float(delta) + 6).truncatingRemainder(dividingBy: 6))
        } else if maxc == g {
            h = 60 * (Float(b - r) / Float(delta) + 2)
        } else {
            h = 60 * (Float(r - g) / Float(delta) + 4)
        }
        out[0] = min(max(h / 2, 0), 179)
        out[1] = min(max(s, 0), 255)
        out[2] = v
    }
}
